lib/Target/SubstraitPB/Import.cpp

```diff
@@ -633,6 +642,115 @@ static mlir::FailureOr<JoinOp> importJoinRel(ImplicitLocOpBuilder builder,
+  filterOp.getCondition().push_back(new Block);
+  Block &conditionBlock = filterOp.getCondition().front();
+  conditionBlock.addArgument(filterOp.getResult().getType(),
+                             filterOp->getLoc());
+
+  // Create condition region.
+  const Expression &expression = rel.post_join_filter();
+  {
+    OpBuilder::InsertionGuard guard(builder);
+    builder.setInsertionPointToEnd(&conditionBlock);
+
+    FailureOr<ExpressionOpInterface> conditionOp =
+        importExpression(builder, expression);
+    if (failed(conditionOp))
+      return failure();
+
+    builder.create<YieldOp>(conditionOp.value()->getResult(0));
+  }
+
+  return {filterOp};
+}
+
+static mlir::FailureOr<RelOpInterface>
+importHashJoinRel(ImplicitLocOpBuilder builder, const Rel &message) {
+  const HashJoinRel &hashJoinRel = message.hash_join();
+  const Rel &leftRel = hashJoinRel.left();
+  const Rel &rightRel = hashJoinRel.right();
+
+  mlir::FailureOr<RelOpInterface> leftOp = importRel(builder, leftRel);
+  mlir::FailureOr<RelOpInterface> rightOp = importRel(builder, rightRel);
+
+  if (failed(leftOp) || failed(rightOp))
+    return failure();
+
+  Value leftVal = leftOp.value()->getResult(0);
+  Value rightVal = rightOp.value()->getResult(0);
+
+  std::optional<JoinType> join_type = static_cast<JoinType>(hashJoinRel.type());
+
+  if (!join_type)
+    return mlir::emitError(builder.getLoc(), "unexpected 'operation' found");
+
+  mlir::FailureOr<HashJoinOp> hashJoinOp =
+      builder.create<HashJoinOp>(leftVal, rightVal, *join_type);
+
+  if (failed(hashJoinOp)) {
+    return failure();
+  }
+
+  hashJoinOp->getCondition().push_back(new Block);
+  Block &conditionBlock = hashJoinOp->getCondition().front();
+  conditionBlock.addArgument(leftVal.getType(), hashJoinOp->getLoc());
+  conditionBlock.addArgument(rightVal.getType(), hashJoinOp->getLoc());
+
+  OpBuilder::InsertionGuard guard(builder);
+  builder.setInsertionPointToEnd(&conditionBlock);
+
+  FailureOr<FieldReferenceOp> leftKeyOp =
+      importFieldReference(builder, hashJoinRel.keys()[0].left());
+  if (failed(leftKeyOp))
+    return failure();
+
+  FailureOr<FieldReferenceOp> rightKeyOp =
+      importFieldReference(builder, hashJoinRel.keys()[0].right());
+  if (failed(rightKeyOp))
+    return failure();
+
+  // Create the comparison operation
+  Value condition;
+  if (hashJoinRel.keys()[0].comparison().has_simple()) {
+    auto simpleComparisonType = static_cast<SimpleComparisonType>(
+        hashJoinRel.keys()[0].comparison().simple());
+
+    Value leftKey = leftKeyOp.value();
+    Value rightKey = rightKeyOp.value();
+
+    condition = builder.create<KeyComparisonOp>(leftKey, rightKey,
+                                                simpleComparisonType);
+  } else {
+    // TODO(trion): Handle custom function if present
+    return mlir::emitError(builder.getLoc(),
+                           "custom comparison functions not yet supported");
```

Comment:
Nit: change to `"custom comparison functions for hash_join not yet supported"` to make it a bit more clear, when one has a very large substrait input file, and it may be a bit hard to decipher what the error message refers to.